bd-session-replay/src/lib.rs

```diff
@@ -0,0 +1,152 @@
+// shared-core - bitdrift's common client/server libraries
+// Copyright Bitdrift, Inc. All rights reserved.
+//
+// Use of this source code is governed by a source available license that can be found in the
+// LICENSE file or at:
+// https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt
+
+#[cfg(test)]
+#[path = "./recorder_test.rs"]
+mod recorder_test;
+
+use bd_runtime::runtime::{session_replay, BoolWatch, ConfigLoader, DurationWatch};
+use bd_shutdown::{ComponentShutdown, ComponentShutdownTrigger};
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::sync::mpsc::{channel, Receiver, Sender};
+use tokio::time::{Instant, Interval};
+
+#[cfg(test)]
+#[ctor::ctor]
+fn test_global_init() {
+  bd_test_helpers::test_global_init();
+}
+
+//
+// Target
+//
+
+pub trait Target {
+  fn capture_wireframes(&self);
+
+  fn take_screenshot(&self);
+}
+
+//
+// Recorder
+//
+
+pub struct Recorder {
+  target: Box<dyn Target + Send + Sync>,
+
+  is_periodic_reporting_enabled: bool,
+  reporting_interval_rate: Duration,
+  reporting_interval: Option<Interval>,
+
+  is_screenshotting_enabled: bool,
+  take_screenshot_rx: Receiver<()>,
+
+  is_periodic_reporting_enabled_flag: BoolWatch<session_replay::PeriodicWireframesEnabledFlag>,
+  reporting_interval_flag: DurationWatch<session_replay::ReportingIntervalFlag>,
+  is_screenshotting_enabled_flag: BoolWatch<session_replay::ScreenshotsEnabledFlag>,
+}
+
+impl Recorder {
+  pub fn new(
+    target: Box<dyn Target + Send + Sync>,
+    runtime_loader: &Arc<ConfigLoader>,
+  ) -> (Self, Sender<()>) {
+    let (take_screenshot_tx, take_screenshot_rx) = channel(2);
+
+    let mut is_periodic_reporting_enabled =
+      session_replay::PeriodicWireframesEnabledFlag::register(runtime_loader).unwrap();
+    let reporting_interval_rate = session_replay::ReportingIntervalFlag::register(runtime_loader)
+      .unwrap()
+      .read_mark_update();
+
+    let mut is_screenshotting_enabled_flag =
+      session_replay::ScreenshotsEnabledFlag::register(runtime_loader).unwrap();
+
+    (
+      Self {
+        target,
+        is_periodic_reporting_enabled: is_periodic_reporting_enabled.read_mark_update(),
+        reporting_interval_rate: reporting_interval_rate.unsigned_abs(),
+        reporting_interval: None,
+        is_screenshotting_enabled: is_screenshotting_enabled_flag.read_mark_update(),
+        take_screenshot_rx,
+        is_periodic_reporting_enabled_flag:
+          session_replay::PeriodicWireframesEnabledFlag::register(runtime_loader).unwrap(),
+        reporting_interval_flag: session_replay::ReportingIntervalFlag::register(runtime_loader)
+          .unwrap(),
+        is_screenshotting_enabled_flag,
+      },
+      take_screenshot_tx,
+    )
+  }
+
+  fn create_interval(interval: Duration, fire_immediately: bool) -> tokio::time::Interval {
```

Comment:
Prefer using time::Duration, then below you can use the interval() and interval_at() helpers in bd_time

Reply:
The two functions below are exactly `interval()` and `interval_at()` unless I am missing something: https://github.com/bitdriftlabs/shared-core/blob/main/bd-time/src/lib.rs#L130-L136